### PR TITLE
[RND-250] Add listener plugin, opensearch as listener, extract AWS from core

### PR DESCRIPTION
**Pull request by bradbanister**

### CODE REVIEW

File: Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/index.ts [hunk 1]
```diff
@@ -3,23 +3,24 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
 
-import { MeadowlarkBackendPlugin } from '@edfi/meadowlark-core';
+import { DocumentStorePlugin, QueryHandlerPlugin } from '@edfi/meadowlark-core';
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/index.ts:6]: Split backend plugin interface into two. One for document store and reference validation (create, update, delete, get by id) and one for handling querys. Can be the same plugin in practice.

File: Meadowlark-js/backends/meadowlark-mongodb-backend/package.json [hunk 2]
```diff
@@ -1,8 +1,8 @@
 {
-  "name": "@edfi/meadowlark-local-mongodb-opensearch",
+  "name": "@edfi/meadowlark-mongodb-backend",
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-mongodb-backend/package.json:2]: "local" became irrelevant

File: Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/DeleteOrchestrator.ts [hunk 3]
```diff
@@ -26,15 +26,15 @@ export async function deleteEntityById({ id, documentInfo, security, traceId }:
       );
 
       if (ownershipResult === 'ERROR') {
-        return { result: 'UNKNOWN_FAILURE' };
+        return { response: 'UNKNOWN_FAILURE' };
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-aws-dynamodb-opensearch/src/DeleteOrchestrator.ts:29]: normalized all the delete/query/get/update/upsert request and response objects

File: Meadowlark-js/backends/meadowlark-opensearch-backend/src/BackendFacade.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  UpdateRequest,
+  UpdateResult,
+  UpsertRequest,
+  UpsertResult,
+} from '@edfi/meadowlark-core';
+import * as QueryOpensearch from './repository/QueryOpensearch';
+import * as UpdateOpensearch from './repository/UpdateOpensearch';
+import { getSharedClient } from './repository/Db';
+
+export async function queryDocuments(request: QueryRequest): Promise<QueryResult> {
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-opensearch-backend/src/BackendFacade.ts:20]: handler for query requests

File: Meadowlark-js/backends/meadowlark-opensearch-backend/src/BackendFacade.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  return QueryOpensearch.queryDocuments(request, await getSharedClient());
+}
+
+export async function afterDeleteDocumentById(request: DeleteRequest, result: DeleteResult) {
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-opensearch-backend/src/BackendFacade.ts:24]: listeners for delete/update/upsert events

File: Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/Db.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/Db.ts:1]: Simplified from old ElasticsearchRepository

File: Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/QueryOpensearch.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/QueryOpensearch.ts:3]: Simplified from old ElasticsearchRepository

File: Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/UpdateOpensearch.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
```
Comment (bradbanister) [inline Meadowlark-js/backends/meadowlark-opensearch-backend/src/repository/UpdateOpensearch.ts:3]: Simplified from old DynamoDbStreamHandler

File: Meadowlark-js/packages/meadowlark-core/src/Logger.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    label: 'Meadowlark',
+  }),
+  winston.format.timestamp({
+    format: 'YYYY-MM-DD HH:mm:SS',
+  }),
+  winston.format.json(),
+);
+
+const offlineFormat = winston.format.combine(
+  winston.format.cli(),
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/Logger.ts:22]: Moved to simpler cli format for offline

File: Meadowlark-js/packages/meadowlark-core/src/handler/Delete.ts [hunk 10]
```diff
@@ -21,58 +24,63 @@ const moduleName = 'Delete';
...
         clientName: jwtStatus.subject,
       },
-      traceId: awsRequestId,
-    });
+      traceId: frontendRequest.traceId,
+    };
 
-    if (result === 'DELETE_SUCCESS') {
-      writeDebugStatusToLog(moduleName, context, 'deleteIt', 204);
+    beforeDeleteDocumentById(request);
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/Delete.ts:62]: Example "before" event publish, here before delete attempt in the document store

File: Meadowlark-js/packages/meadowlark-core/src/handler/Delete.ts [hunk 11]
```diff
@@ -21,58 +24,63 @@ const moduleName = 'Delete';
...
+    const result: DeleteResult = await getDocumentStore().deleteDocumentById(request);
+    afterDeleteDocumentById(request, result);
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/Delete.ts:64]: Example "after" event publish, here before delete attempt in the document store

File: Meadowlark-js/packages/meadowlark-core/src/handler/DescriptorLoader.ts [hunk 12]
```diff
@@ -35,6 +35,21 @@ type DescriptorDocument = {
   description: string;
 };
 
+/** Returns a new object */
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/DescriptorLoader.ts:38]: these were in Utility, never ended up being used anywhere else

File: Meadowlark-js/packages/meadowlark-core/src/handler/FrontendRequest.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/FrontendRequest.ts:3]: Object interface from frontend incoming API request to meadowlark-core

File: Meadowlark-js/packages/meadowlark-core/src/handler/FrontendResponse.ts [hunk 14]
```diff
@@ -3,8 +3,8 @@
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/FrontendResponse.ts:5]: Object interface from meadowlark-core API result to frontend

File: Meadowlark-js/packages/meadowlark-core/src/handler/Get.ts [hunk 15]
```diff
@@ -19,42 +19,43 @@ const moduleName = 'Get';
  *
  * Determines whether request is "get all", "get by id", or a query, and forwards to the appropriate handler
  */
-export async function getResolver(event: APIGatewayProxyEvent, context: Context): Promise<APIGatewayProxyResult> {
+export async function getResolver(frontendRequest: FrontendRequest): Promise<FrontendResponse> {
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/handler/Get.ts:22]: Example of removal of AWS from meadowlark-core

File: Meadowlark-js/packages/meadowlark-core/src/message/GetResult.ts [hunk 16]
```diff
@@ -4,7 +4,8 @@
 // See the LICENSE and NOTICES files in the project root for more information.
 
 export type GetResult = {
-  result: 'ERROR' | 'NOT_FOUND' | 'SUCCESS' | 'AUTHORIZATION_FAILURE' | 'LEGACY_AUTHORIZATION_FAILURE';
-  documents: Array<object>;
+  response: 'GET_SUCCESS' | 'GET_FAILURE_NOT_EXISTS' | 'GET_FAILURE_AUTHORIZATION' | 'UNKNOWN_FAILURE';
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/message/GetResult.ts:7]: Example of normalizing backend request/result objects

File: Meadowlark-js/packages/meadowlark-core/src/plugin/PluginLoader.ts [hunk 17]
```diff
@@ -2,35 +2,87 @@
...
+      loadedQueryHandler = await loadQueryHandler(process.env.QUERY_HANDLER_PLUGIN);
+    } catch (e) {
+      const message = e instanceof Error ? e.message : 'unknown';
+      Logger.error(`Unable to load query handler plugin "${process.env.QUERY_HANDLER_PLUGIN}". Error was ${message}`, null);
+      throw e;
+    }
+  }
+
+  // If this works out, we'll discover listeners dynamically - no need to specify
+  if (process.env.LISTENER1_PLUGIN != null) {
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/plugin/PluginLoader.ts:61]: For now, hardcoded the possibility of one or two listener plugins via env variables. No limit in practice though.

File: Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+export const BEFORE_GET_DOCUMENT_BY_ID = Symbol('BEFORE_GET_DOCUMENT_BY_ID');
+export const AFTER_GET_DOCUMENT_BY_ID = Symbol('AFTER_GET_DOCUMENT_BY_ID');
+export const BEFORE_UPDATE_DOCUMENT_BY_ID = Symbol('BEFORE_UPDATE_DOCUMENT_BY_ID');
+export const AFTER_UPDATE_DOCUMENT_BY_ID = Symbol('AFTER_UPDATE_DOCUMENT_BY_ID');
+export const BEFORE_DELETE_DOCUMENT_BY_ID = Symbol('BEFORE_DELETE_DOCUMENT_BY_ID');
+export const AFTER_DELETE_DOCUMENT_BY_ID = Symbol('AFTER_DELETE_DOCUMENT_BY_ID');
+export const BEFORE_QUERY_DOCUMENTS = Symbol('BEFORE_QUERY_DOCUMENTS');
+export const AFTER_QUERY_DOCUMENTS = Symbol('AFTER_QUERY_DOCUMENTS');
+
+export const emitter: Emittery = new Emittery();
```
Comment (bradbanister) [inline Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts:24]: The in-process pub/sub library used for our listener plugins. Supports async/await, very popular, skimmed the codebase.

File: Meadowlark-js/services/meadowlark-aws-lambda/src/handler/MeadowlarkConverter.ts [hunk 19]
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+/* eslint-disable-next-line import/no-unresolved */
+import type { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
+import type { FrontendRequest, FrontendResponse } from '@edfi/meadowlark-core';
+
```
Comment (bradbanister) [inline Meadowlark-js/services/meadowlark-aws-lambda/src/handler/MeadowlarkConverter.ts:9]: Simple translation from AWS request/response to meadowlark-core's

File: Meadowlark-js/test/http/local.mongo.with.simple.opensearch.http [hunk 20]
```diff
@@ -83,8 +83,26 @@ DELETE http://localhost:3000/local/v3.3b/ed-fi/academicWeeks/20325050be22032deae
 
 
 
+######################### Query
```
Comment (bradbanister) [inline Meadowlark-js/test/http/local.mongo.with.simple.opensearch.http:86]: Manual tests for Opensearch queries

File: Meadowlark-js/packages/meadowlark-core/src/handler/Upsert.ts [hunk 21]
```diff
@@ -3,105 +3,115 @@
...
-    });
-    if (result === 'INSERT_SUCCESS') {
-      writeDebugStatusToLog(moduleName, context, 'create', 201);
-      const location = `${event.path}${event.path.endsWith('/') ? '' : '/'}${resourceId}`;
+      traceId: frontendRequest.traceId,
+    };
+
+    beforeUpsertDocument(request);
+    const result: UpsertResult = await getDocumentStore().upsertDocument(request);
+    afterUpsertDocument(request, result);
```
Comment (stephenfuqua) [inline Meadowlark-js/packages/meadowlark-core/src/handler/Upsert.ts:83]: If change data capture becomes a hassle in some way, we could also use this notification process to log to Kafka.

File: Meadowlark-js/packages/meadowlark-core/src/metaed/LoadMetaEd.ts [hunk 22]
```diff
@@ -57,7 +57,7 @@ function metaEdConfigurationFrom(metaEdProjectMetadata: MetaEdProjectMetadata[])
 
 // TODO: should probably include project version in key
```
Comment (stephenfuqua) [inline Meadowlark-js/packages/meadowlark-core/src/metaed/LoadMetaEd.ts:58]:
```suggestion
// TODO in RND-260: should probably include project version in key
```

File: Meadowlark-js/packages/meadowlark-core/src/plugin/PluginLoader.ts (hunk 17 above)
Comment (stephenfuqua) [reply Meadowlark-js/packages/meadowlark-core/src/plugin/PluginLoader.ts:61]: Seeing this, and having seen the code above, I now better understand. This isn't (for now) a "subscribe here" polling mechanism, although a listener plugin might provide that. The listener is in fact running inline with the main process. Which is very interesting and worthwhile.

File: Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+export const BEFORE_UPSERT_DOCUMENT = Symbol('BEFORE_UPSERT_DOCUMENT');
+export const AFTER_UPSERT_DOCUMENT = Symbol('AFTER_UPSERT_DOCUMENT');
+export const BEFORE_GET_DOCUMENT_BY_ID = Symbol('BEFORE_GET_DOCUMENT_BY_ID');
+export const AFTER_GET_DOCUMENT_BY_ID = Symbol('AFTER_GET_DOCUMENT_BY_ID');
+export const BEFORE_UPDATE_DOCUMENT_BY_ID = Symbol('BEFORE_UPDATE_DOCUMENT_BY_ID');
+export const AFTER_UPDATE_DOCUMENT_BY_ID = Symbol('AFTER_UPDATE_DOCUMENT_BY_ID');
+export const BEFORE_DELETE_DOCUMENT_BY_ID = Symbol('BEFORE_DELETE_DOCUMENT_BY_ID');
+export const AFTER_DELETE_DOCUMENT_BY_ID = Symbol('AFTER_DELETE_DOCUMENT_BY_ID');
+export const BEFORE_QUERY_DOCUMENTS = Symbol('BEFORE_QUERY_DOCUMENTS');
+export const AFTER_QUERY_DOCUMENTS = Symbol('AFTER_QUERY_DOCUMENTS');
```
Comment (stephenfuqua) [inline Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts:22]: Can listeners filter messages by symbol? For example, to only react to the "AFTER_*" symbols.

**Review by stephenfuqua — COMMENTED** [review]
Submitting comments, next need to fetch and run it all.

File: Meadowlark-js/services/meadowlark-aws-lambda/.env.example [hunk 24]
```diff
@@ -29,5 +29,12 @@ ES_PASSWORD=admin
...
-BACKEND_PLUGIN_NAME=@edfi/meadowlark-local-mongodb-opensearch
+#DOCUMENT_STORE_PLUGIN=@edfi/meadowlark-aws-dynamodb-opensearch
+DOCUMENT_STORE_PLUGIN=@edfi/meadowlark-mongodb-backend
+
+#QUERY_HANDLER_PLUGIN=@edfi/meadowlark-aws-dynamodb-opensearch
+QUERY_HANDLER_PLUGIN=@edfi/meadowlark-opensearch-backend
+
+#LISTENER1_PLUGIN=@edfi/meadowlark-opensearch-backend
+
+#LISTENER2_PLUGIN=@edfi/meadowlark-test-listener
```
Comment (stephenfuqua) [inline Meadowlark-js/services/meadowlark-aws-lambda/.env.example:40]: Would it make sense to move this to the `Meadowlark-js` directory?

Comment (stephenfuqua) [reply Meadowlark-js/services/meadowlark-aws-lambda/.env.example:40]: Well, I see that this wouldn't make sense. However, I find it very odd to have to dig this deep into the directory structure just to start things up.

File: Meadowlark-js/services/meadowlark-aws-lambda/serverless.yml [hunk 25]
```diff
@@ -42,7 +42,7 @@ custom:
 
 provider:
   name: aws
-  runtime: nodejs16.x
+  runtime: nodejs14.x
   stage: ${opt:stage, 'local'}
   timeout: 10
```
Comment (stephenfuqua) [inline Meadowlark-js/services/meadowlark-aws-lambda/serverless.yml:47]: What do you think about trying to have multiple serverless config files? e.g. one with dynamo and one without.

**Comment by stephenfuqua** [comment]
Random observation...

image

File: Meadowlark-js/packages/meadowlark-core/src/handler/Upsert.ts (hunk 21 above)
Comment (bradbanister) [reply Meadowlark-js/packages/meadowlark-core/src/handler/Upsert.ts:83]: 👍

File: Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts (hunk 23 above)
Comment (bradbanister) [reply Meadowlark-js/packages/meadowlark-core/src/plugin/listener/Publish.ts:22]: Yes, but via the registration methods in Subscribe.ts. The symbols are basically Emittery "topics" that we aren't exposing directly.

File: Meadowlark-js/services/meadowlark-aws-lambda/serverless.yml (hunk 25 above)
Comment (bradbanister) [reply Meadowlark-js/services/meadowlark-aws-lambda/serverless.yml:47]: Quite likely. In the back of my mind, I've been thinking about how we might specify AWS Lambda but with different permutations of AWS services. Maybe similar configurations (like always using DocumentDB, but differing on using managed Kafka) go in the same Meadowlark service package with alternate serverless.yaml files, but less similar ones get their own package, for example a crazy direct-to-S3-for-storage one in something like "meadowlark-aws-lambda-s3-athena".

**Review by stephenfuqua — APPROVED** [review]

**Pull request merged**